File: rules_sonatype/src/main/scala/org/jetbrains/bazel/sonatype/SonatypeClient.scala
package org.jetbrains.bazel.sonatype

import org.apache.http.auth.{AuthScope, UsernamePasswordCredentials}
import org.apache.http.impl.client.BasicCredentialsProvider
import org.jetbrains.bazel.sonatype.SonatypeClient.{ActivityMonitor, CreateStageResponse, StageTransitionRequest, StagingActivity, StagingProfile, StagingProfileResponse, StagingRepositoryProfile}
import org.jetbrains.bazel.sonatype.SonatypeException._
import org.sonatype.spice.zapper.ParametersBuilder
import org.sonatype.spice.zapper.client.hc4.Hc4ClientBuilder
import org.sonatype.spice.zapper.fs.DirectoryIOSource
import wvlet.airframe.control.{Control, ResultClass, Retry}
import wvlet.airframe.http.HttpHeader.MediaType
import wvlet.airframe.http.HttpMessage.{Request, Response}
import wvlet.airframe.http._
import wvlet.airframe.http.client.{URLConnectionClient, URLConnectionClientConfig}
import wvlet.log.LogSupport

import java.io.IOException
import java.nio.charset.StandardCharsets
import java.util.Base64
import java.util.concurrent.TimeUnit
import scala.concurrent.duration.Duration
import scala.util.{Failure, Success, Try}

class SonatypeClient(
    repositoryUrl: String,
    username: String,
    password: String,
    timeoutMillis: Int = 60 * 60 * 1000
) extends AutoCloseable
    with LogSupport {
  private val base64Credentials =
    Base64.getEncoder.encodeToString(s"$username:$password".getBytes(StandardCharsets.UTF_8))

  lazy val repoUri: String = {
    def repoBase(url: String) = if (url.endsWith("/")) url.dropRight(1) else url
    val url                   = repoBase(repositoryUrl)
    url
  }
  private val pathPrefix = {
    new java.net.URL(repoUri).getPath
  }

  private object SonatypeClientBackend extends HttpClientBackend {
    def newSyncClient(serverAddress: String, clientConfig: HttpClientConfig): HttpSyncClient[Request, Response] = {
      new URLConnectionClient(
        ServerAddress(serverAddress),
        URLConnectionClientConfig(
          requestFilter = clientConfig.requestFilter,
          retryContext = clientConfig.retryContext,
          codecFactory = clientConfig.codecFactory,
          readTimeout = Duration(timeoutMillis, TimeUnit.MILLISECONDS)
        )
      )
    }
  }

  private val clientConfig = HttpClientConfig(SonatypeClientBackend)
    // airframe-http will retry the request several times within this timeout duration.
    .withRetryContext { context =>
      // For individual REST calls, use a normal jittering
      context
        .withMaxRetry(15)
        .withJitter(initialIntervalMillis = 1500, maxIntervalMillis = 30000)
    }
    .withRequestFilter { request =>
      request.withContentTypeJson
        .withAccept(MediaType.ApplicationJson)
        .withHeader(HttpHeader.Authorization, s"Basic $base64Credentials")
    }

  private val httpClient = clientConfig.newSyncClient(repoUri)

  // Create stage is not idempotent, so we just need to wait for a long time without retry
  private val httpClientForCreateStage =
    clientConfig
      .withRetryContext(_.noRetry)
      .newSyncClient(repoUri)

  override def close(): Unit = {
    Control.closeResources(httpClient, httpClientForCreateStage)
  }
  def stagingRepositoryProfiles: Seq[StagingRepositoryProfile] = {
    info("Reading staging repository profiles...")
    val result =
      httpClient.get[Map[String, Seq[StagingRepositoryProfile]]](s"$pathPrefix/staging/profile_repositories")
    result.getOrElse("data", Seq.empty)
  }

  def stagingRepository(repositoryId: String): String = {
    info(s"Searching for repository $repositoryId ...")
    httpClient.get[String](s"$pathPrefix/staging/repository/$repositoryId")
  }

  def stagingProfiles: Seq[StagingProfile] = {
    info("Reading staging profiles...")
    val result = httpClient.get[StagingProfileResponse](s"$pathPrefix/staging/profiles")
    result.data
  }

  def createStage(profile: StagingProfile, description: String): StagingRepositoryProfile = {
    info(s"Creating a staging repository in profile ${profile.name} with a description key: $description")
    val ret = httpClientForCreateStage.postOps[Map[String, Map[String, String]], CreateStageResponse](
      s"$pathPrefix/staging/profiles/${profile.id}/start",
      Map("data" -> Map("description" -> description))
    )
    // Extract created staging repository ids
    val repo = StagingRepositoryProfile(
      profile.id,
      profile.name,
      "open",
      repositoryId = ret.data.stagedRepositoryId,
      ret.data.description
    )
    info(s"Created successfully: ${repo.repositoryId}")
    repo
  }

  private val monitor = new ActivityMonitor()

  /** backoff retry (max 15 sec. / each http request) until the timeout reaches (upto 60 min by default)
    */
  private val retryer = {
    val maxInterval  = 15000
    val initInterval = 3000
    // init * (multiplier ^ n) = max
    // n = log(max / init) / log(multiplier)
    val retryCountUntilMaxInterval = (math.log(maxInterval.toDouble / initInterval) / math.log(1.5)).toInt.max(1)
    val numRetry                   = (timeoutMillis.toFloat / maxInterval).ceil.toInt
    Retry.withBackOff(
      maxRetry = retryCountUntilMaxInterval + numRetry,
      initialIntervalMillis = initInterval,
      maxIntervalMillis = maxInterval
    )
  }

  private def waitForStageCompletion(
      taskName: String,
      repo: StagingRepositoryProfile,
      terminationCond: StagingActivity => Boolean
  ): StagingRepositoryProfile = {
    retryer
      .beforeRetry { ctx =>
        ctx.lastError match {
          case SonatypeException(STAGE_IN_PROGRESS, msg) =>
            info(f"$msg Waiting for ${ctx.nextWaitMillis / 1000.0}%.2f sec.")
          case _ =>
            warn(
              f"[${ctx.retryCount}/${ctx.maxRetry}] Execution failed: ${ctx.lastError.getMessage}. Retrying in ${ctx.nextWaitMillis / 1000.0}%.2f sec."
            )
        }
      }
      .withResultClassifier[Option[StagingActivity]] {
        case Some(activity) if terminationCond(activity) =>
          info(s"[$taskName] Finished successfully")
          ResultClass.Succeeded
        case Some(activity) if activity.containsError =>
          error(s"[$taskName] Failed")
          activity.reportFailure()
          ResultClass.nonRetryableFailure(SonatypeException(STAGE_FAILURE, s"Failed to $taskName the repository."))
        case _ =>
          ResultClass.retryableFailure(SonatypeException(STAGE_IN_PROGRESS, s"The $taskName stage is in progress."))
      }
      .run {
        val activities = activitiesOf(repo)
        monitor.report(activities)
        activities.lastOption
      }

    repo
  }

  def closeStage(currentProfile: StagingProfile, repo: StagingRepositoryProfile): StagingRepositoryProfile = {
    info(s"Closing staging repository $repo")
    val ret = httpClient.postOps[Map[String, StageTransitionRequest], Response](
      s"$pathPrefix/staging/profiles/${repo.profileId}/finish",
      newStageTransitionRequest(currentProfile, repo)
    )
    if (ret.statusCode != HttpStatus.Created_201.code) {
      throw SonatypeException(STAGE_FAILURE, s"Failed to close the repository. [${ret.status}]: ${ret.contentString}")
    }
    waitForStageCompletion(
      "close",
      repo,
      terminationCond = {
        _.isCloseSucceeded(repo.repositoryId)
      }
    ).toClosed
  }

  def promoteStage(currentProfile: StagingProfile, repo: StagingRepositoryProfile): StagingRepositoryProfile = {
    info(s"Promoting staging repository $repo")
    val ret = httpClient.postOps[Map[String, StageTransitionRequest], Response](
      s"$pathPrefix/staging/profiles/${repo.profileId}/promote",
      newStageTransitionRequest(currentProfile, repo)
    )
    if (ret.statusCode != HttpStatus.Created_201.code) {
      throw SonatypeException(STAGE_FAILURE, s"Failed to promote the repository. [${ret.status}]: ${ret.contentString}")
    }

    waitForStageCompletion("promote", repo, terminationCond = { _.isReleaseSucceeded(repo.repositoryId) })
  }

  def dropStage(currentProfile: StagingProfile, repo: StagingRepositoryProfile): Response = {
    info(s"Dropping staging repository $repo")
    val ret = httpClient.postOps[Map[String, StageTransitionRequest], Response](
      s"$pathPrefix/staging/profiles/${repo.profileId}/drop",
      newStageTransitionRequest(currentProfile, repo)
    )
    if (ret.statusCode != HttpStatus.Created_201.code) {
      throw SonatypeException(STAGE_FAILURE, s"Failed to drop the repository. [${ret.status}]: ${ret.contentString}")
    }
    ret
  }

  private def newStageTransitionRequest(
      currentProfile: StagingProfile,
      repo: StagingRepositoryProfile
  ): Map[String, StageTransitionRequest] = {
    Map(
      "data" -> StageTransitionRequest(
        stagedRepositoryId = repo.repositoryId,
        targetRepositoryId = currentProfile.repositoryTargetId,
        description = repo.description
      )
    )
  }

  def activitiesOf(r: StagingRepositoryProfile): Seq[StagingActivity] = {
    debug(s"Checking activity logs of ${r.repositoryId} ...")
    httpClient.get[Seq[StagingActivity]](s"$pathPrefix/staging/repository/${r.repositoryId}/activity")
  }

  def uploadBundle(coordinateGroupId: String, remoteUrl: String, getDeployables: () => DirectoryIOSource): Unit = {
    retryer
      .retryOn {
        case e: IOException if e.getMessage.contains("400 Bad Request") =>
          Retry.nonRetryableFailure(
            SonatypeException(
              BUNDLE_UPLOAD_FAILURE,
              s"Bundle upload failed. Probably a previously uploaded bundle remains. Run sonatypeClean or sonatypeDropAll first: ${e.getMessage}"
            )
          )
        case e: SigningException =>
          Retry.nonRetryableFailure(e)
      }
      .withResultClassifier[Try[_]] {
        case Success(_) => ResultClass.Succeeded
        case Failure(x) => ResultClass.Failed(isRetryable = false, x)
      }
      .run {
        val parameters = ParametersBuilder.defaults().build()
        // Adding a trailing slash is necessary upload a bundle file to a proper location:
        val endpoint      = s"$remoteUrl/"
        val clientBuilder = new Hc4ClientBuilder(parameters, endpoint)

        val credentialProvider = new BasicCredentialsProvider()
        val usernamePasswordCredentials =
          new UsernamePasswordCredentials(username, password)

        credentialProvider.setCredentials(AuthScope.ANY, usernamePasswordCredentials)

        clientBuilder.withPreemptiveRealm(credentialProvider)
        val deployables = getDeployables()

        val client = clientBuilder.build()
        val tried = Try {
          info(s"Uploading bundle $coordinateGroupId to $endpoint")
          client.upload(deployables)
          info(s"Finished bundle upload: $coordinateGroupId")
        }
        client.close()
        tried
      }
  }
}

object SonatypeClient extends LogSupport {

  case class StagingProfileResponse(data: Seq[StagingProfile] = Seq.empty)

  /** Staging profile is the information associated to a Sonatype account.
    */
  case class StagingProfile(id: String, name: String, repositoryTargetId: String)

  /** Staging repository profile has an id of deployed artifact and the current staging state.
    */
  case class StagingRepositoryProfile(
      profileId: String,
      profileName: String,
      `type`: String,
      repositoryId: String,
      description: String
  ) {
    def stagingType: String = `type`

    override def toString =
      s"[$repositoryId] status:$stagingType, profile:$profileName($profileId) description: $description"
    def isOpen: Boolean = stagingType == "open"
    def isClosed: Boolean = stagingType == "closed"
    def isReleased: Boolean = stagingType == "released"

    def toClosed: StagingRepositoryProfile = copy(`type` = "closed")
    def toDropped: StagingRepositoryProfile = copy(`type` = "dropped")
    def toReleased: StagingRepositoryProfile = copy(`type` = "released")

    def deployUrl(repositoryUrl: String): String = s"$repositoryUrl/staging/deployByRepositoryId/$repositoryId"
  }

  case class CreateStageResponse(
      data: StagingRepositoryRef
  )
  //noinspection ScalaWeakerAccess
  case class StagingRepositoryRef(
      stagedRepositoryId: String,
      description: String
  )

  case class StageTransitionRequest(
      stagedRepositoryId: String,
      targetRepositoryId: String,
      description: String
  )

  //noinspection ScalaWeakerAccess
  case class Prop(name: String, value: String)

  /** ActivityEvent is an evaluation result (e.g., checksum, signature check, etc.)
    * of a rule defined in a StagingActivity ruleset
    */
  case class ActivityEvent(timestamp: String, name: String, severity: Int, properties: Seq[Prop]) {
    lazy val map: Map[String, String] = properties.map(x => x.name -> x.value).toMap
    def ruleType: String = map.getOrElse("typeId", "other")
    def isFailure: Boolean = name == "ruleFailed"

    override def toString: String = {
      s"-event -- timestamp:$timestamp, name:$name, severity:$severity, ${properties.map(p => s"${p.name}:${p.value}").mkString(", ")}"
    }

    def showProgress(useErrorLog: Boolean = false): Unit = {
      val props = {
        val front =
          if (map.contains("typeId"))
            Seq(map("typeId"))
          else
            Seq.empty
        front ++ map.filter(_._1 != "typeId").map(p => s"${p._1}:${p._2}")
      }
      val messageLine = props.mkString(", ")
      val name_s      = name.replaceAll("rule(s)?", "")
      val message     = f"$name_s%10s: $messageLine"
      if (useErrorLog)
        logger.error(message)
      else
        logger.info(message)
    }
  }

  class ActivityMonitor {
    private var reportedActivities = Set.empty[String]
    private var reportedEvents     = Set.empty[ActivityEvent]

    def report(stagingActivities: Seq[StagingActivity]): Unit = {
      for (sa <- stagingActivities) {
        if (!reportedActivities.contains(sa.started)) {
          logger.info(sa.activityLog)
          reportedActivities += sa.started
        }
        for (ae <- sa.events if !reportedEvents.contains(ae)) {
          ae.showProgress(useErrorLog = false)
          reportedEvents += ae
        }
      }
    }
  }

  /** Staging activity is an action to the staged repository
    * @param name activity name, e.g. open, close, promote, etc.
    */
  case class StagingActivity(name: String, started: String, stopped: String, events: Seq[ActivityEvent]) {
    override def toString: String = {
      val b = Seq.newBuilder[String]
      b += activityLog
      for (e <- events)
        b += s" ${e.toString}"
      b.result().mkString("\n")
    }

    def activityLog: String = {
      val b = Seq.newBuilder[String]
      b += s"Activity name:$name"
      b += s"started:$started"
      if (stopped.nonEmpty) {
        b += s"stopped:$stopped"
      }
      b.result().mkString(", ")
    }

    def showProgress(): Unit = {
      logger.info(activityLog)
      val hasError = containsError
      for (e <- suppressEvaluateLog) {
        e.showProgress(hasError)
      }
    }

    private def suppressEvaluateLog: Seq[ActivityEvent] = {
      val in     = events.toIndexedSeq
      var cursor = 0
      val b      = Seq.newBuilder[ActivityEvent]
      while (cursor < in.size) {
        val current = in(cursor)
        if (cursor < in.size - 1) {
          val next = in(cursor + 1)
          if (current.name == "ruleEvaluate" && current.ruleType == next.ruleType) {
            // skip
          } else {
            b += current
          }
        }
        cursor += 1
      }
      b.result()
    }

    def containsError: Boolean = events.exists(_.severity != 0)

    def failureReport: Seq[ActivityEvent] = suppressEvaluateLog.filter(_.isFailure)

    def reportFailure(): Unit = {
      logger.error(activityLog)
      for (e <- failureReport) {
        e.showProgress(useErrorLog = true)
      }
    }

    def isReleaseSucceeded(repositoryId: String): Boolean = {
      events
        .find(_.name == "repositoryReleased")
        .exists(_.map.getOrElse("id", "") == repositoryId)
    }

    def isCloseSucceeded(repositoryId: String): Boolean = {
      events
        .find(_.name == "repositoryClosed")
        .exists(_.map.getOrElse("id", "") == repositoryId)
    }
  }
}

File: sonatype_kt/src/main/kotlin/org/jetbrains/bazel/sonatype/SonatypeClient.kt
package org.jetbrains.bazel.sonatype

import org.apache.http.auth.AuthScope
import org.apache.http.auth.UsernamePasswordCredentials
import org.apache.http.client.methods.*
import org.apache.http.impl.client.BasicCredentialsProvider
import org.apache.http.impl.client.CloseableHttpClient
import org.apache.http.impl.client.HttpClients
import org.apache.http.entity.StringEntity
import org.apache.http.util.EntityUtils
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import org.sonatype.spice.zapper.client.hc4.Hc4ClientBuilder
import org.sonatype.spice.zapper.ParametersBuilder
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import java.io.IOException
import java.net.URL
import java.nio.charset.StandardCharsets
import java.util.*
import java.util.concurrent.TimeUnit

class SonatypeClient(
  repositoryUrl: String,
  private val username: String,
  private val password: String,
  private val timeoutMillis: Int = 60 * 60 * 1000
) : AutoCloseable {

  val LOG: Logger = LogManager.getLogger(SonatypeClient::class.java)

  private val base64Credentials: String =
    Base64.getEncoder().encodeToString("$username:$password".toByteArray(StandardCharsets.UTF_8))

  private val repoUri: String = repositoryUrl.trimEnd('/')

  private val pathPrefix: String = URL(repoUri).path

  private val mapper = jacksonObjectMapper()

  private val httpClient: CloseableHttpClient

  init {
    val credentialsProvider = BasicCredentialsProvider()
    credentialsProvider.setCredentials(AuthScope.ANY, UsernamePasswordCredentials(username, password))

    httpClient = HttpClients.custom()
      .setDefaultCredentialsProvider(credentialsProvider)
      .setConnectionTimeToLive(timeoutMillis.toLong(), TimeUnit.MILLISECONDS)
      .build()
  }

  override fun close() {
    httpClient.close()
  }

  // Function to execute a GET request and parse the response into a specific type
  private inline fun <reified T> getRequest(url: String): T {
    val request = HttpGet(url).apply {
      addHeader("Accept", "application/json")
      addHeader("Authorization", "Basic $base64Credentials")
    }
    httpClient.execute(request).use { response ->
      val status = response.statusLine.statusCode
      val entity = response.entity ?: throw IOException("Empty response body")
      val content = EntityUtils.toString(entity, StandardCharsets.UTF_8)
      if (status in 200..299) {
        return mapper.readValue(content)
      } else {
        throw IOException("GET request failed with status $status: $content")
      }
    }
  }

  // Function to execute a POST request and parse the response into a specific type
  private inline fun <reified Req, reified Res> postRequest(url: String, body: Req): Res {
    val request = HttpPost(url).apply {
      addHeader("Content-Type", "application/json")
      addHeader("Accept", "application/json")
      addHeader("Authorization", "Basic $base64Credentials")
      entity = StringEntity(mapper.writeValueAsString(body), StandardCharsets.UTF_8)
    }
    httpClient.execute(request).use { response ->
      val status = response.statusLine.statusCode
      val entity = response.entity ?: throw IOException("Empty response body")
      val content = EntityUtils.toString(entity, StandardCharsets.UTF_8)
      if (status in 200..299) {
        return mapper.readValue(content)
      } else {
        throw IOException("POST request failed with status $status: $content")
      }
    }
  }

  // Data classes for responses
  data class StagingProfileResponse(val data: List<StagingProfile> = emptyList())
  data class StagingProfile(
    val id: String,
    val name: String,
    val repositoryTargetId: String
  )
  data class StagingRepositoryProfile(
    val profileId: String,
    val profileName: String,
    val type: String,
    val repositoryId: String,
    val description: String
  ) {
    fun isOpen() = type == "open"
    fun isClosed() = type == "closed"
    fun isReleased() = type == "released"

    fun toClosed(): StagingRepositoryProfile = copy(type = "closed")
    fun toDropped(): StagingRepositoryProfile = copy(type = "dropped")
    fun toReleased(): StagingRepositoryProfile = copy(type = "released")

    fun deployUrl(repositoryUrl: String): String = "$repositoryUrl/staging/deployByRepositoryId/$repositoryId"

    override fun toString(): String =
      "[$repositoryId] status:$type, profile:$profileName($profileId) description: $description"
  }

  data class CreateStageResponse(
    val data: StagingRepositoryRef
  )
  data class StagingRepositoryRef(
    val stagedRepositoryId: String,
    val description: String
  )
  data class StageTransitionRequest(
    val stagedRepositoryId: String,
    val targetRepositoryId: String,
    val description: String
  )

  data class CloseStageResponse(val statusCode: Int, val content: String)
  data class PromoteStageResponse(val statusCode: Int, val content: String)
  data class ResponseDropStage(val statusCode: Int, val content: String)

  // Implement stagingRepositoryProfiles
  fun stagingRepositoryProfiles(): List<StagingRepositoryProfile> {
    LOG.info("Reading staging repository profiles...")
    val url = "$repoUri$pathPrefix/staging/profile_repositories"
    return getRequest<StagingRepositoryProfileResponseWrapper>(url).data
  }

  // Wrapper to match expected JSON
  data class StagingRepositoryProfileResponseWrapper(val data: List<StagingRepositoryProfile> = emptyList())

  // Implement stagingRepository
  fun stagingRepository(repositoryId: String): String {
    LOG.info("Searching for repository $repositoryId ...")
    val url = "$repoUri$pathPrefix/staging/repository/$repositoryId"
    return getRequest<String>(url)
  }

  // Implement stagingProfiles
  fun stagingProfiles(): List<StagingProfile> {
    LOG.info("Reading staging profiles...")
    val url = "$repoUri$pathPrefix/staging/profiles"
    return getRequest<StagingProfileResponse>(url).data
  }

  // Implement createStage
  fun createStage(profile: StagingProfile, description: String): StagingRepositoryProfile {
    LOG.info("Creating a staging repository in profile ${profile.name} with a description key: $description")
    val url = "$repoUri$pathPrefix/staging/profiles/${profile.id}/start"
    val requestBody = mapOf("data" to StageTransitionRequest(
      stagedRepositoryId = "",
      targetRepositoryId = profile.repositoryTargetId,
      description = description
    ))
    val response: CreateStageResponse = postRequest(url, requestBody)
    val repo = StagingRepositoryProfile(
      profileId = profile.id,
      profileName = profile.name,
      type = "open",
      repositoryId = response.data.stagedRepositoryId,
      description = response.data.description
    )
    LOG.info("Created successfully: ${repo.repositoryId}")
    return repo
  }

  // Implement closeStage
  fun closeStage(profile: StagingProfile, repo: StagingRepositoryProfile): StagingRepositoryProfile {
    LOG.info("Closing staging repository $repo")
    val url = "$repoUri$pathPrefix/staging/profiles/${repo.profileId}/finish"
    val requestBody = newStageTransitionRequest(profile, repo)
    val response: CloseStageResponse = postRequest(url, requestBody)
    if (response.statusCode != 201) { // Assuming 201 is Created
      throw SonatypeException(SonatypeErrorCode.STAGE_FAILURE, "Failed to close the repository. [${response.statusCode}]: ${response.content}")
    }
    return waitForStageCompletion(
      "close",
      repo,
      terminationCond = { activity -> activity.isCloseSucceeded(repo.repositoryId) }
    ).toClosed()
  }

  // Implement promoteStage
  fun promoteStage(profile: StagingProfile, repo: StagingRepositoryProfile): StagingRepositoryProfile {
    LOG.info("Promoting staging repository $repo")
    val url = "$repoUri$pathPrefix/staging/profiles/${repo.profileId}/promote"
    val requestBody = newStageTransitionRequest(profile, repo)
    val response: PromoteStageResponse = postRequest(url, requestBody)
    if (response.statusCode != 201) { // Assuming 201 is Created
      throw SonatypeException(SonatypeErrorCode.STAGE_FAILURE, "Failed to promote the repository. [${response.statusCode}]: ${response.content}")
    }
    return waitForStageCompletion(
      "promote",
      repo,
      terminationCond = { activity -> activity.isReleaseSucceeded(repo.repositoryId) }
    )
  }

  // Implement dropStage
  fun dropStage(profile: StagingProfile, repo: StagingRepositoryProfile): ResponseDropStage {
    LOG.info("Dropping staging repository $repo")
    val url = "$repoUri$pathPrefix/staging/profiles/${repo.profileId}/drop"
    val requestBody = newStageTransitionRequest(profile, repo)
    val response: ResponseDropStage = postRequest(url, requestBody)
    if (response.statusCode != 201) { // Assuming 201 is Created
      throw SonatypeException(SonatypeErrorCode.STAGE_FAILURE, "Failed to drop the repository. [${response.statusCode}]: ${response.content}")
    }
    return response
  }

  // Implement uploadBundle
  fun uploadBundle(coordinateGroupId: String, remoteUrl: String, getDeployables: () -> DirectoryIOSourceMaven) {
    try {
      LOG.info("Uploading bundle $coordinateGroupId to $remoteUrl")
      val parameters = ParametersBuilder.defaults().build()
      val clientBuilder = Hc4ClientBuilder(parameters, remoteUrl)

      val credentialProvider = org.apache.http.impl.client.BasicCredentialsProvider()
      val creds = UsernamePasswordCredentials(username, password)
      credentialProvider.setCredentials(AuthScope.ANY, creds)
      clientBuilder.withPreemptiveRealm(credentialProvider)

      val client = clientBuilder.build()
      val deployables = getDeployables()

      client.upload(deployables)
      LOG.info("Finished bundle upload: $coordinateGroupId")
      client.close()
    } catch (e: Exception) {
      LOG.error("Bundle upload failed: ${e.message}")
      throw SonatypeException(SonatypeErrorCode.BUNDLE_UPLOAD_FAILURE, "Bundle upload failed: ${e.message}")
    }
  }

  // Implement waitForStageCompletion
  fun waitForStageCompletion(
    taskName: String,
    repo: StagingRepositoryProfile,
    terminationCond: (StagingActivity) -> Boolean
  ): StagingRepositoryProfile {
    val maxInterval = 15000L
    val initInterval = 3000L
    val multiplier = 1.5
    val retryCountUntilMaxInterval = (Math.log(maxInterval.toDouble() / initInterval) / Math.log(multiplier)).toInt().coerceAtLeast(1)
    val numRetry = Math.ceil(timeoutMillis.toDouble() / maxInterval).toInt()
    val totalRetries = retryCountUntilMaxInterval + numRetry

    val monitor = ActivityMonitor()

    for (retry in 1..totalRetries) {
      try {
        val activities = activitiesOf(repo)
        monitor.report(activities)
        val lastActivity = activities.lastOrNull()
        if (lastActivity != null) {
          when {
            terminationCond(lastActivity) -> {
              LOG.info("[$taskName] Finished successfully")
              return repo
            }
            lastActivity.containsError() -> {
              LOG.error("[$taskName] Failed")
              lastActivity.reportFailure()
              throw SonatypeException(SonatypeErrorCode.STAGE_FAILURE, "Failed to $taskName the repository.")
            }
            else -> {
              // Wait and retry
              val waitTime = (initInterval * Math.pow(multiplier, retry.toDouble())).toLong().coerceAtMost(maxInterval)
              LOG.info("Waiting for ${waitTime / 1000.0} seconds before retrying...")
              Thread.sleep(waitTime)
            }
          }
        }
      } catch (e: SonatypeException) {
        LOG.error("Error during $taskName: ${e.message}")
        throw e
      } catch (e: Exception) {
        LOG.warn("[$retry/$totalRetries] Execution failed: ${e.message}. Retrying in ${initInterval / 1000.0} seconds.")
        val waitTime = (initInterval * Math.pow(multiplier, retry.toDouble())).toLong().coerceAtMost(maxInterval)
        Thread.sleep(waitTime)
      }
    }
    throw SonatypeException(SonatypeErrorCode.STAGE_IN_PROGRESS, "The $taskName stage is in progress.")
  }

  // Implement activitiesOf
  fun activitiesOf(repo: StagingRepositoryProfile): List<StagingActivity> {
    LOG.debug("Checking activity logs of ${repo.repositoryId} ...")
    val url = "$repoUri$pathPrefix/staging/repository/${repo.repositoryId}/activity"
    return getRequest(url)
  }

  // Implement helper methods
  private fun newStageTransitionRequest(profile: StagingProfile, repo: StagingRepositoryProfile): Map<String, StageTransitionRequest> {
    return mapOf(
      "data" to StageTransitionRequest(
        stagedRepositoryId = repo.repositoryId,
        targetRepositoryId = profile.repositoryTargetId,
        description = repo.description
      )
    )
  }

  // ActivityMonitor and related classes
  class ActivityMonitor {
    private val reportedActivities = mutableSetOf<String>()
    private val reportedEvents = mutableSetOf<ActivityEvent>()

    private val LOG: Logger = LogManager.getLogger(ActivityMonitor::class.java)

    fun report(stagingActivities: List<StagingActivity>) {
      for (sa in stagingActivities) {
        if (!reportedActivities.contains(sa.started)) {
          LOG.info(sa.activityLog())
          reportedActivities.add(sa.started)
        }
        for (ae in sa.events) {
          if (!reportedEvents.contains(ae)) {
            ae.showProgress(false)
            reportedEvents.add(ae)
          }
        }
      }
    }
  }

  // Define ActivityEvent and StagingActivity
  data class ActivityEvent(
    val timestamp: String,
    val name: String,
    val severity: Int,
    val properties: List<Prop>
  ) {
    val LOG = LogManager.getLogger(ActivityEvent::class.java)
    val map: Map<String, String> = properties.associate { it.name to it.value }
    fun ruleType(): String = map["typeId"] ?: "other"
    fun isFailure(): Boolean = name == "ruleFailed"

    override fun toString(): String =
      "-event -- timestamp:$timestamp, name:$name, severity:$severity, " +
        properties.joinToString(", ") { "${it.name}:${it.value}" }

    fun showProgress(useErrorLog: Boolean) {
      val props = mutableListOf<String>()
      if (map.containsKey("typeId")) {
        props.add(map["typeId"]!!)
      }
      props += map.filterKeys { it != "typeId" }.map { "${it.key}:${it.value}" }
      val messageLine = props.joinToString(", ")
      val name_s = name.replace(Regex("rule(s)?"), "")
      val message = String.format("%10s: %s", name_s, messageLine)
      if (useErrorLog) {
        LOG.error(message)
      } else {
        LOG.info(message)
      }
    }
  }

  data class Prop(val name: String, val value: String)

  data class StagingActivity(
    val name: String,
    val started: String,
    val stopped: String,
    val events: List<ActivityEvent>
  ) {
    val LOG = LogManager.getLogger(StagingActivity::class.java)

    override fun toString(): String {
      val builder = StringBuilder()
      builder.append(activityLog())
      for (e in events) {
        builder.append("\n ").append(e.toString())
      }
      return builder.toString()
    }

    fun activityLog(): String {
      val builder = StringBuilder()
      builder.append("Activity name:$name")
      builder.append(", started:$started")
      if (stopped.isNotEmpty()) {
        builder.append(", stopped:$stopped")
      }
      return builder.toString()
    }

    fun showProgress() {
      LOG.info(activityLog())
      val hasError = containsError()
      for (e in suppressEvaluateLog()) {
        e.showProgress(hasError)
      }
    }

    private fun suppressEvaluateLog(): List<ActivityEvent> {
      val inEvents = events
      val b = mutableListOf<ActivityEvent>()
      var cursor = 0
      while (cursor < inEvents.size) {
        val current = inEvents[cursor]
        if (cursor < inEvents.size - 1) {
          val next = inEvents[cursor + 1]
          if (current.name == "ruleEvaluate" && current.ruleType() == next.ruleType()) {
            // skip
          } else {
            b.add(current)
          }
        } else {
          b.add(current)
        }
        cursor += 1
      }
      return b
    }

    fun containsError(): Boolean = events.any { it.severity != 0 }

    fun failureReport(): List<ActivityEvent> = suppressEvaluateLog().filter { it.isFailure() }

    fun reportFailure() {
      LOG.error(activityLog())
      for (e in failureReport()) {
        e.showProgress(true)
      }
    }

    fun isReleaseSucceeded(repositoryId: String): Boolean {
      return events.find { it.name == "repositoryReleased" }
        ?.map?.getOrDefault("id", "") == repositoryId
    }

    fun isCloseSucceeded(repositoryId: String): Boolean {
      return events.find { it.name == "repositoryClosed" }
        ?.map?.getOrDefault("id", "") == repositoryId
    }
  }
}
